src/core/lib/event_engine/windows/windows_endpoint.cc

```diff
@@ -0,0 +1,333 @@
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#ifdef GPR_WINDOWS
+
+#include "absl/cleanup/cleanup.h"
+#include "absl/functional/any_invocable.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_format.h"
+
+#include <grpc/event_engine/memory_allocator.h>
+#include <grpc/support/log_windows.h>
+
+#include "src/core/lib/event_engine/trace.h"
+#include "src/core/lib/event_engine/windows/resolved_address.h"
+#include "src/core/lib/event_engine/windows/windows_endpoint.h"
+#include "src/core/lib/gprpp/debug_location.h"
+#include "src/core/lib/gprpp/status_helper.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// DO NOT SUBMIT(hork): we should be careful about the changes to the error
+// return types, they were previously all Unavailable.
+// TODO(hork): The previous implementation required internal ref counting. Add
+// this when it becomes necessary.
+// TODO(hork): The previous implementation required a 2-phase shutdown. Add this
+// when it becomes necessary.
+
+namespace {
+constexpr int kDefaultTargetReadSize = 8192;
+constexpr int kMaxWSABUFCount = 16;
+
+void AbortOnEvent(absl::Status) {
+  GPR_ASSERT(false &&
+             "INTERNAL ERROR: Asked to handle read/write event with an invalid "
+             "callback");
+}
+
+absl::Status WSAErrorToStatusWithMessage(
+    int wsa_error, absl::string_view custom_message,
+    const grpc_core::DebugLocation& location) {
+  // See
+  // https://learn.microsoft.com/en-us/windows/win32/api/Winsock2/nf-winsock2-wsasend
+  // https://learn.microsoft.com/en-us/windows/win32/api/Winsock2/nf-winsock2-wsarecv
+  char* wsa_message = gpr_format_message(wsa_error);
+  std::string message;
+  if (!custom_message.empty()) {
+    std::string message = absl::StrCat(wsa_message, ": ", custom_message);
```

Comment:
remove the `std::string` here - we don't want to declare a new variable and then immediately throw it away